kernel/src/engine/arrow_expression.rs

```diff
@@ -366,6 +364,193 @@ fn evaluate_expression(
     }
 }
 
+// Apply a schema to an array. The array _must_ be a `StructArray`. Returns a `RecordBatch where the
+// names of fields, nullable, and metadata in the struct have been transformed to match those in
+// schema specified by `schema`
+fn apply_schema(array: &dyn Array, schema: &DataType) -> DeltaResult<RecordBatch> {
+    let DataType::Struct(struct_schema) = schema else {
+        return Err(Error::generic(
+            "apply_schema at top-level must be passed a struct schema",
+        ))?;
+    };
+    let applied = apply_schema_to_struct(array, struct_schema)?;
+    Ok(applied.into())
+}
+
+// helper to transform an arrow field+col into the specified target type. If `rename` is specified
+// the field will be renamed to the contained `str`.
+fn transform_field_and_col(
+    arrow_field: &Arc<ArrowField>,
+    arrow_col: Arc<dyn Array>,
+    target_type: &DataType,
+    nullable: bool,
+    rename: Option<&str>,
+    metadata: Option<HashMap<String, String>>,
+) -> DeltaResult<(ArrowField, Arc<dyn Array>)> {
+    let transformed_col = apply_schema_to(&arrow_col, target_type)?.unwrap_or(arrow_col);
```

Comment:
clarifying: if we fail to apply schema to `arrow_col` we fail (propagate error with `?`) but in the `None` case we have already validated that the `arrow_col` matches `target_type` and so it's safe to just let transformed = arrow_col? maybe comment to help since there's a lot packed in here?